### PR TITLE
JAMES-3693 Add healthcheck test for case redis sentinel

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+        return parameterContext.getParameter().getType() == RedisSentinelCluster.class;
+    }
+
+    @Override
+    public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+        return new RedisSentinelCluster(redisMasterReplicaContainer, redisSentinelContainer);
+    }
+
+    private static String createRedisURI(List<GenericContainer> containers) {
```
Comment (quantranhong1999) [inline backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java:195]: createRedisSentinelURI

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+import io.lettuce.core.ReadFrom;
+import scala.Function2;
+import scala.jdk.javaapi.OptionConverters;
+
+public class RedisSentinelExtension implements GuiceModuleTestExtension {
+    public static final int SENTINEL_PORT = 26379;
+
+    public static class RedisMasterReplicaContainer extends ArrayList<GenericContainer> {
+        public RedisMasterReplicaContainer(Collection<? extends GenericContainer> c) {
```
Comment (quantranhong1999) [inline backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java:60]: `RedisMasterReplicaContainer` name should reflect containing replica containers too, no?

**Review by quantranhong1999 — APPROVED** [review]

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            super(c);
+        }
+
+        public void pauseOne() {
```
Comment (quantranhong1999) [inline backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java:64]: What is `One`? Should it be `pauseMasterNode` I guess?

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java (hunk 2 above)
Comment (hungphan227) [reply backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java]: yeah it also contains replicas

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java (hunk 3 above)
Comment (Arsnael) [reply backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java]: It seems he is pausing the first node out of 3 to me..?

**Comment by Arsnael** [comment]
Can we have a docker-compose somewhere of james running with redis sentinel?

**Comment by hungphan227** [comment]
> Can we have a docker-compose somewhere of james running with redis sentinel?

@vttranlina is working on docker compose for sentinel for Mauricius deployment and I guess he is doing well. If you want docker compose for james, taking it from tung would be better

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
> > Can we have a docker-compose somewhere of james running with redis sentinel?
> 
> @vttranlina is working on docker compose for sentinel for Mauricius deployment and I guess he is doing well. If you want docker compose for james, taking it from tung would be better

Currently, I have not already a docker-compose that combines Sentinel with James, but here is my docker-compose for experiment only Sentinel:
https://github.com/vttranlina/james-project/tree/redis_lab/backends-common/redis/redis-sentinel-docker-compose

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelHealthCheckTest.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ************************************************************** */
+
+package org.apache.james.backends.redis
+
+import java.util.concurrent.TimeUnit
+
+import org.apache.james.backends.redis.RedisMasterReplicaExtension.RedisMasterReplicaContainer
```
Comment (vttranlina) [inline backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelHealthCheckTest.scala:24]: un-used import

**Review by vttranlina — COMMENTED** [review]
I am concerned that a only healthCheck test may not be sufficient. The healthCheck will make sense when James uses Sentinel. Therefore, we need to write an integration test for any feature, that using RedisSentinelExtension.

**Comment by quantranhong1999** [comment]
> I am concerned that a only healthCheck test may not be sufficient. The healthCheck will make sense when James uses Sentinel. Therefore, we need to write an integration test for any feature, that using RedisSentinelExtension.

Very valid point to me.

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        }
+    }
+
+    public static class RedisSentinelContainerList extends ArrayList<GenericContainer> {
+        public RedisSentinelContainerList(Collection<? extends GenericContainer> c) {
+            super(c);
+        }
+
+        public MasterReplicaRedisConfiguration getRedisConfiguration() {
+            return MasterReplicaRedisConfiguration.from(ImmutableList.of(createRedisSentinelURI(this.stream().toList().subList(0,1)))
```
Comment (vttranlina) [inline backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisSentinelExtension.java:86]: Q: 
we have 3 sentinel nodes
Why you use only one here?

**Comment by vttranlina** [comment]
I contributed add more test cases for [RedisRateLimiterWithSentinelTest](https://github.com/apache/james-project/pull/2358/commits/b130942fd4a7f2b4fe28b326955dc459e4d949a1)

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
Updated: 
I pushed a docker-compose sample in another pr https://github.com/apache/james-project/pull/2362

**Pull request merged**